packages/lib-algorand/index.ts

```diff
@@ -0,0 +1,89 @@
+import { DfnsApiClient, DfnsError } from '@dfns/sdk'
+import { GetWalletResponse, GenerateSignatureResponse } from '@dfns/sdk/types/wallets'
+
+import { encodeObj, Transaction, EncodedSignedTransaction, decodeObj } from 'algosdk'
+export type DfnsWalletOptions = {
+  walletId: string
+  dfnsClient: DfnsApiClient
+}
+
+const hexToBuffer = (hex: string): Buffer => {
+  return Buffer.from(hex.replace(/^0x/, ''), 'hex')
+}
+
+export const singleSigTransactionToEncoded = (tx: Transaction): EncodedSignedTransaction =>{
+  return {
+    txn: tx.get_obj_for_encoding()
+  }
+}
+
+type WalletMetadata = GetWalletResponse
+
+const assertSigned = (res: GenerateSignatureResponse) => {
+  if (res.status === 'Failed') {
+    throw new DfnsError(-1, 'signing failed', res)
+  } else if (res.status !== 'Signed') {
+    throw new DfnsError(
+      -1,
+      'cannot complete signing synchronously because this wallet action requires policy approval',
+      res
+    )
+  }
+}
+
+export class DfnsWallet {
+  private readonly dfnsClient: DfnsApiClient
+
+  private constructor(private metadata: WalletMetadata, options: DfnsWalletOptions) {
+    this.dfnsClient = options.dfnsClient
+  }
+
+  public static async init(options: DfnsWalletOptions) {
+    const { walletId, dfnsClient } = options
+    const res = await dfnsClient.wallets.getWallet({ walletId })
+
+    if (res.status !== 'Active') {
+      throw new DfnsError(-1, 'wallet not active', { walletId, status: res.status })
+    }
+
+    const { scheme, curve } = res.signingKey
+    if (scheme !== 'EdDSA') {
+      throw new DfnsError(-1, 'key scheme is not EdDSA', { walletId, scheme })
+    }
+    if (curve !== 'ed25519') {
+      throw new DfnsError(-1, 'key curve is not ed25519', { walletId, curve })
+    }
+
+    if (res.network !== 'Algorand' && res.network !== 'AlgorandTestnet') {
+      throw new DfnsError(-1, 'wallet is not bound to Algorand or AlgorandTestnet', {
+        walletId,
+        network: res.network,
+      })
+    }
+
+    return new DfnsWallet(res, options)
+  }
+
+  public get address(): string {
+    return this.metadata.address!
+
+  }
+
+  // sign single signature transaction
+  public async signTransaction(transaction: EncodedSignedTransaction): Promise<EncodedSignedTransaction> {
+    const res = await this.dfnsClient.wallets.generateSignature({
+      walletId: this.metadata.id,
+      body: {
+        kind: 'Transaction',
+        transaction: `0x${Buffer.from(encodeObj(transaction)).toString('hex')}`,
+      },
+    })
+
+    assertSigned(res)
+    if (!res.signature?.encoded) {
```

Comment:
should check `!res.signedData` instead because we will try to decode signed data.